component/src/governance/view.rs

```diff
@@ -166,6 +118,13 @@ pub trait StateWriteExt: StateWrite {
     /// Store a new proposal with a new proposal id.
     async fn new_proposal(&mut self, proposal: &Proposal) -> Result<u64> {
         let proposal_id = self.next_proposal_id().await?;
+        if proposal_id != proposal.id {
+            return Err(anyhow::anyhow!(
+                "proposal id {} does not match next proposal id {}",
+                proposal.id,
+                proposal_id
+            ));
+        }
```

Comment:
Proposals now self-contain their ID, so we need to check that this matches here before modifying the state.